llvm/compiler-course/backend/chistov_a_combination/chistov_a_backend.cpp

```diff
@@ -0,0 +1,143 @@
+    for (auto &MBB : MF) {
+      llvm::SmallVector<llvm::MachineInstr *, 8> InstrsToRemove;
+      for (auto &MI : MBB) {
+        if (shouldTransform(MI)) {
+          performTransform(MBB, MI, InstrsToRemove);
+          Changed = true;
+        }
+      }
+      for (auto *I : InstrsToRemove)
+        I->eraseFromParent();
+    }
+  }
+
+  bool isLogicOpcode(unsigned Opcode) const {
+    return LogicToAVX.contains(Opcode);
+  }
+
+  unsigned getAVXOpcode(unsigned Opcode) const {
+    auto It = LogicToAVX.find(Opcode);
+    return It != LogicToAVX.end() ? It->second : 0;
+  }
+
+  void replaceWithAVX(llvm::MachineBasicBlock &MBB, llvm::MachineInstr &MI,
+                      llvm::SmallVectorImpl<llvm::MachineInstr *> &ToRemove) {
+    unsigned NewOpc = getAVXOpcode(MI.getOpcode());
+    if (!NewOpc)
+      return;
+
+    llvm::Register DestReg = MI.getOperand(0).getReg();
+    llvm::Register SrcReg1 = MI.getOperand(1).getReg();
+    llvm::Register SrcReg2 = MI.getOperand(2).getReg();
+
+    llvm::BuildMI(MBB, MI, MI.getDebugLoc(), TII->get(NewOpc), DestReg)
+        .addReg(SrcReg1)
+        .addReg(SrcReg2);
+    ToRemove.push_back(&MI);
+    Changed = true;
+  }
+
+private: // optimizing passes
+  void optimizeLogicOperations(llvm::MachineFunction &MF) {
+    OptimizationPredicate shouldTransform = [&](llvm::MachineInstr &MI) {
+      return isLogicOpcode(MI.getOpcode());
+    };
+
+    OptimizationTransform transformInstr =
+        [&](llvm::MachineBasicBlock &MBB, llvm::MachineInstr &MI,
+            llvm::SmallVectorImpl<llvm::MachineInstr *> &ToRemove) {
+          replaceWithAVX(MBB, MI, ToRemove);
+        };
+
+    applyOptimizations(MF, shouldTransform, transformInstr);
+  }
+
+  void combineLogicOperations(llvm::MachineFunction &MF) {
+    OptimizationPredicate shouldTransform = [&](llvm::MachineInstr &MI) {
+      return isLogicOpcode(MI.getOpcode());
+    };
+
+    OptimizationTransform transformInstr =
+        [&](llvm::MachineBasicBlock &MBB, llvm::MachineInstr &MI,
+            llvm::SmallVectorImpl<llvm::MachineInstr *> &InstrsToRemove) {
+          llvm::Register IntermediateReg = MI.getOperand(1).getReg();
+          auto *FirstInstr = MRI->getUniqueVRegDef(IntermediateReg);
+          if (!FirstInstr || !MRI->hasOneUse(IntermediateReg))
+            return;
+
+          unsigned FirstOpcode = FirstInstr->getOpcode();
+          if (!isLogicOpcode(FirstOpcode))
+            return;
+
+          unsigned AVXFirstOp = getAVXOpcode(FirstOpcode);
+          unsigned AVXSecondOp = getAVXOpcode(MI.getOpcode());
+          if (!AVXFirstOp || !AVXSecondOp)
+            return;
+
+          replaceWithAVX(MBB, *FirstInstr, InstrsToRemove);
+          replaceWithAVX(MBB, MI, InstrsToRemove);
+        };
+
+    applyOptimizations(MF, shouldTransform, transformInstr);
+  }
+
+public:
+  static char ID;
+  X86LogicOptPass() : llvm::MachineFunctionPass(ID) {}
+
+  bool runOnMachineFunction(llvm::MachineFunction &MF) override {
```

Comment:
`Changed`variable needs to be reset with `false` in the beginning of the function if it is the class member